### PR TITLE
PART IX: Pixar Mesh write translator/utilities refactoring 

### PR DESCRIPTION
**Pull request by HamedSabri-adsk**
### PART IX
- Start using `UsdMayaWriteUtil::SetAttribute` utilities in all writers

```
cameraWriter
instancerWriter
jointWriter
nurbsCurveWriter
nurbsSurfaceWriter
particleWriter
```

### CODE REVIEW

File: lib/mayaUsd/fileio/utils/writeUtil.h [hunk 1]
```diff
@@ -330,6 +330,16 @@ struct UsdMayaWriteUtil
...
+    template <typename T>
+    static bool SetAttribute(const UsdAttribute& usdAttr,
+                             const T& value,
+                             const UsdTimeCode& usdTime,
+                             UsdUtilsSparseValueWriter *valueWriter)
+    {
+        return valueWriter ?
+               valueWriter->SetAttribute(usdAttr, VtValue(value), usdTime) :
+               usdAttr.Set(value, usdTime);
+    }
```
[the remark this answers is not in the document]
Comment (HamedSabri-adsk) [reply lib/mayaUsd/fileio/utils/writeUtil.h]: @mattyjams  see 062ac5a9b3ec945033a26ae88700c3bb90b56434

Fair enough, I was originally just using the available code and not thinking to refactor them. As you rightly mentioned, having different parameter order is confusing.

The two suggested versions makes sense with both UsdTimeCode and UsdUtilsSparseValueWriter as optional parameters, however I don't believe they would work in cases where we want to pass UsdUtilsSparseValueWriter but not UsdTimeCode:

e.g

jointWriteUtil.cpp

`UsdMayaWriteUtil::SetAttribute(bindingAPI.CreateGeomBindTransformAttr(), &geomBindTransform, valueWriter);`

```
error C2665: 'pxrInternal_v0_20__pxrReserved__::UsdMayaWriteUtil::SetAttribute4': none of the 2 overloads could convert all the argument types

note: could be 'bool pxrInternal_v0_20__pxrReserved__::UsdMayaWriteUtil::SetAttribute4<pxrInternal_v0_20__pxrReserved__::GfMatrix4d>(const pxrInternal_v0_20__pxrReserved__::UsdAttribute &,T *,const pxrInternal_v0_20__pxrReserved__::UsdTimeCode,pxrInternal_v0_20__pxrReserved__::UsdUtilsSparseValueWriter *)'
```

In the original code, UsdUtilsSparseValueWriter was passed as the third option ( not optional ) which would also work with the scenario above it if we want it to be optional. 

We can still keep the order as the original code and with your suggested versions, the new two version would look like this:

```
template <typename T>
static bool SetAttribute(
        const UsdAttribute& attr,
        const T& value,
        UsdUtilsSparseValueWriter* valueWriter = nullptr,
        const UsdTimeCode time = UsdTimeCode::Default())
{
    return valueWriter ?
        valueWriter->SetAttribute(attr, VtValue(value), time) :
        attr.Set(value, time);
}

template <typename T>
static bool SetAttribute(
        const UsdAttribute& attr,
        T* value,
        UsdUtilsSparseValueWriter* valueWriter = nullptr,
        const UsdTimeCode time = UsdTimeCode::Default())
{
    return valueWriter ?
        valueWriter->SetAttribute(attr, VtValue::Take(*value), time) :
        attr.Set(*value, time);
}
```

[the remark this answers is not in the document]
Comment (mattyjams) [reply lib/mayaUsd/fileio/utils/writeUtil.h]: It's probably splitting hairs at this point, but it makes a little more sense to me that `valueWriter` be the final optional parameter.

Setting an attribute conceptually is:
author attribute `attr` using value `value` at time code `time`

Those parameters all affect **what** is being authored. `valueWriter` on the other hand affects **how** it's being authored.

But with that, you're right that you can't have an invocation like this:
```C++
UsdMayaWriteUtil::SetAttribute(
    bindingAPI.CreateGeomBindTransformAttr(),
    &geomBindTransform,
    valueWriter);
```

You would instead have to also provide the time code:
```C++
UsdMayaWriteUtil::SetAttribute(
    bindingAPI.CreateGeomBindTransformAttr(),
    &geomBindTransform,
    UsdTimeCode::Default(),
    valueWriter);
```

If you're going so far as to be using a `valueWriter`, you have specific needs as to how the attribute is authored, so it doesn't seem like an additional burden that you must also pass in a time code.

[the remark this answers is not in the document]
Comment (HamedSabri-adsk) [reply lib/mayaUsd/fileio/utils/writeUtil.h]: `Those parameters all affect what is being authored. valueWriter on the other hand affects how it's being authored.`
@mattyjams Makes sense.  see d891b7e2bc88343d6a8d3f7f6d41ee462f9d0d1a

[the remark this answers is not in the document]
Comment (mattyjams) [reply lib/mayaUsd/fileio/utils/writeUtil.h]: Great, thanks!

**Pull request merged**